### DISCUSSION TITLE
Load Testing results

### DISCUSSION THREAD

**Discussion by tjhenry** [WVAviator/myco-kv#9] [General]
I was trying to crash it :) Mem usage looking nice and linear as expected. The 'too many open files' is (probably) either the os or python not closing file descriptors (network connections) fast enough, but not totally sure. I started slow, then ramped up speed by an order or magnitude until there was no delay being added. Not a bug, just wanted to share.

<img width="1896" alt="Screenshot 2023-07-09 at 3 06 11 PM" src="https://github.com/WVAviator/myco-kv/assets/1296004/dd3c32a3-6009-4c7b-ad13-28d9827500f2">

<img width="1896" alt="MEM Screenshot" src="https://github.com/WVAviator/myco-kv/assets/1296004/60386e5f-e323-4522-a220-04b7c7df84c1">

**Comment by WVAviator** [comment]
Nice! In it's current state, it's most significant performance bottleneck would probably be persistence - currently to restore the db state it simply reads from the same log file that it writes to for each command. So every restart it repeats every command you've ever run.

I've been trying to think of ways to optimize this, and it seems like a CAP theroem problem. I could just serialize the entire db, but that's a big operation that would need to occur with every command to ensure persistence (slow, weakened availability). I could also run worker threads that periodically serialize. This keeps it fast and available but you sacrifice persistence if the db or server crashes and all writes are lost since the last save.

One thought I had was to still use the log file method, but read and apply the logs in reverse. Then, if an operation is requested on a key, all following operations on that same key can be ignored (since you'd have the most up-to-date value for that key). From all that, a brand new log file could be created that only consists of the required operations and ignores irrelevant older changes. This will provide a benefit only if a large number of operations on a small set of keys is performed (if all operations are done on unique keys, this will be just as slow as the current method). This idea also gets complex when considering how I'm storing and working with nested keys.

Thanks for running a stress test though, I might have to try doing that myself as well! I should probably also reread DDIA and try to better understand persistence. Sorry to go off on a tangent.
